File: tobas_setup_assistant/src/tobas_setup_assistant/setting_widgets/observer/eskf.py
from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ...setup_assistant import SetupAssistant

from overrides import override
from PyQt5.QtWidgets import QCheckBox, QVBoxLayout

from tobas_rqt_tools.messages import q_error_named

from .base import BaseObserver


class ErrorStateKalmanFilter(BaseObserver):
    NAME = "Error State Kalman Filter"
    PACKAGE_NAME = "state_estimation_eskf"
    ABST_TEXT = (
        "The Error State Kalman Filter (ESKF) is an advanced variant of the Kalman Filter, "
        "tailored for systems with non-linear dynamics. "
        "Unlike the traditional Kalman Filter, which directly estimates the system's state, "
        "the ESKF focuses on estimating the error in the state. "
        "This approach allows for more effective handling of non-linear relationships "
        "between the system state and measurements. "
        "The ESKF operates by linearizing these non-linearities around a nominal state. "
        "It's particularly useful in applications like navigation and tracking, "
        "where precision in estimating orientation and position is crucial, "
        "such as in Inertial Navigation Systems and GPS technology. "
        "The ESKF's blend of accuracy and computational efficiency "
        "makes it a valuable tool in complex engineering tasks."
    )

    def __init__(self, main: SetupAssistant) -> None:
        super().__init__(main)

        self._param_rows = QVBoxLayout()
        self._rows.addLayout(self._param_rows)

        self._do_acc_bias_estimation = QCheckBox(text="Do Accelerometer Bias Estimation")
        self._do_acc_bias_estimation.setChecked(False)
        self._param_rows.addWidget(self._do_acc_bias_estimation)

        self._do_gyro_bias_estimation = QCheckBox(text="Do Gyroscope Bias Estimation")
        self._do_gyro_bias_estimation.setChecked(True)
        self._param_rows.addWidget(self._do_gyro_bias_estimation)

        self._do_gravity_estimation = QCheckBox(text="Do Gravity Estimation")
        self._do_gravity_estimation.setChecked(True)
        self._param_rows.addWidget(self._do_gravity_estimation)

    @override
    def is_valid(self) -> bool:
        # 絶対位置が取得できないとダメ
        no_gps = not self._main.gps.equipped()
        no_odom = not self._main.odometry.equipped()
        if no_gps and no_odom:
            q_error_named(
                self._main, self.NAME, "Absolute position connot be observed. Please review the sensor settings."
            )
            return False

        return True

    @override
    def dump_settings(self) -> dict:
        res = dict()
        for i in range(self._param_rows.count()):
            ckb: QCheckBox = self._param_rows.itemAt(i).widget()
            res[ckb.text()] = ckb.isChecked()
        return res

    @override
    def load_settings(self, data: dict) -> None:
        for i in range(self._param_rows.count()):
            ckb: QCheckBox = self._param_rows.itemAt(i).widget()
            ckb.setChecked(data[ckb.text()])

    @override
    def static_parameters(self) -> dict:
        return {
            "use_barometer": False,  # TODO: 選択できるように
            "use_gps": self._main.gps.equipped(),
            "do_acc_bias_estimation": self._do_acc_bias_estimation.isChecked(),
            "do_gyro_bias_estimation": self._do_gyro_bias_estimation.isChecked(),
            "do_gravity_estimation": self._do_gravity_estimation.isChecked(),
            "imu_offset": self._main.imu.offset.get(),
            "barometer_offset": self._main.barometer.offset.get(),
            "gps_offset": self._main.gps.offset.get(),
        }

File: tobas_gui/tobas_setup_assistant/src/setting_widgets/observer/eskf.cpp
#include <tobas_yaml_tools/convert/eigen.hpp>
#include <tobas_yaml_tools/convert/qstring.hpp>
#include <tobas_qt_tools/message.hpp>

#include "tobas_setup_assistant/setting_tabs/observer/eskf.hpp"

namespace gui
{
namespace setup_assistant
{
ErrorStateKalmanFilterWidget::ErrorStateKalmanFilterWidget(
  const RobotInfo& robot,
  const IMUWidget* imu,
  const BarometerWidget* baro,
  const GPSWidget* gps)
  : robot_(robot), imu_(imu), baro_(baro), gps_(gps)
{
  const auto rows = new QVBoxLayout();
  setLayout(rows);

  do_acc_bias_estimation_ = new QCheckBox("Do Accelerometer Bias Estimation");
  do_acc_bias_estimation_->setChecked(false);
  rows->addWidget(do_acc_bias_estimation_);

  do_gyro_bias_estimation_ = new QCheckBox("Do Gyroscope Bias Estimation");
  do_gyro_bias_estimation_->setChecked(true);
  rows->addWidget(do_gyro_bias_estimation_);

  do_mag_hard_bias_estimation_ = new QCheckBox("Do Magnetometer Hard-Iron Bias Estimation");
  do_mag_hard_bias_estimation_->setChecked(true);
  rows->addWidget(do_mag_hard_bias_estimation_);

  do_mag_soft_bias_estimation_ = new QCheckBox("Do Magnetometer Soft-Iron Bias Estimation");
  do_mag_soft_bias_estimation_->setChecked(true);
  rows->addWidget(do_mag_soft_bias_estimation_);

  do_grav_estimation_ = new QCheckBox("Do Gravity Estimation");
  do_grav_estimation_->setChecked(true);
  rows->addWidget(do_grav_estimation_);

  rows->addStretch();
}

const char* ErrorStateKalmanFilterWidget::name() const
{
  return "Error State Kalman Filter";
}

const char* ErrorStateKalmanFilterWidget::description() const
{
  return "The Error State Kalman Filter (ESKF) is an advanced variant of the Kalman Filter, "
         "tailored for systems with non-linear dynamics. "
         "Unlike the traditional Kalman Filter, which directly estimates the system's state, "
         "the ESKF focuses on estimating the error in the state. "
         "This approach allows for more effective handling of non-linear relationships "
         "between the system state and measurements. "
         "The ESKF operates by linearizing these non-linearities around a nominal state. "
         "It's particularly useful in applications like navigation and tracking, "
         "where precision in estimating orientation and position is crucial, "
         "such as in Inertial Navigation Systems and GPS technology. "
         "The ESKF's blend of accuracy and computational efficiency "
         "makes it a valuable tool in complex engineering tasks.";
}

QString ErrorStateKalmanFilterWidget::observerPackage() const
{
  return "tobas_eskf";
}

QString ErrorStateKalmanFilterWidget::pluginName() const
{
  return "ObserverNode";
}

YAML::Node ErrorStateKalmanFilterWidget::staticParams() const
{
  YAML::Node node(YAML::NodeType::Map);

  node["frame_id"] = robot_.tree().getRootName();
  node["use_barometer"] = false;  // TODO: 選択できるように
  node["use_gps"] = gps_->equipped();
  node["do_acc_bias_estimation"] = do_acc_bias_estimation_->isChecked();
  node["do_gyro_bias_estimation"] = do_gyro_bias_estimation_->isChecked();
  node["do_mag_hard_bias_estimation"] = do_mag_hard_bias_estimation_->isChecked();
  node["do_mag_soft_bias_estimation"] = do_mag_soft_bias_estimation_->isChecked();
  node["do_gravity_estimation"] = do_grav_estimation_->isChecked();
  node["imu_offset"] = imu_->offset();
  node["barometer_offset"] = baro_->offset();
  node["gps_offset"] = gps_->offset();

  return node;
}

YAML::Node ErrorStateKalmanFilterWidget::dump() const
{
  YAML::Node node(YAML::NodeType::Map);

  node[do_acc_bias_estimation_->text()] = do_acc_bias_estimation_->isChecked();
  node[do_gyro_bias_estimation_->text()] = do_gyro_bias_estimation_->isChecked();
  node[do_mag_hard_bias_estimation_->text()] = do_mag_hard_bias_estimation_->isChecked();
  node[do_mag_soft_bias_estimation_->text()] = do_mag_soft_bias_estimation_->isChecked();
  node[do_grav_estimation_->text()] = do_grav_estimation_->isChecked();

  return node;
}

void ErrorStateKalmanFilterWidget::load(const YAML::Node& node)
{
  do_acc_bias_estimation_->setChecked(node[do_acc_bias_estimation_->text()].as<bool>());
  do_gyro_bias_estimation_->setChecked(node[do_gyro_bias_estimation_->text()].as<bool>());
  do_mag_hard_bias_estimation_->setChecked(node[do_mag_hard_bias_estimation_->text()].as<bool>());
  do_mag_soft_bias_estimation_->setChecked(node[do_mag_soft_bias_estimation_->text()].as<bool>());
  do_grav_estimation_->setChecked(node[do_grav_estimation_->text()].as<bool>());
}

bool ErrorStateKalmanFilterWidget::isValid()
{
  // 絶対位置が取得できることを確認
  if (!gps_->equipped())
  {
    qt::qErrorBox(this, "Absolute position connot be observed. Please review the sensor settings.");
    return false;
  }

  return true;
}
}  // namespace setup_assistant
}  // namespace gui
